### PR TITLE
Instrumented code to measure execution time of the various layers

### LINKED ISSUE #1032: Add an easy way to measure time spent in layers
We have [PerformanceReports](https://github.com/json-api-dotnet/PerformanceReports) to run load tests and verify scalability, but these are black-box tests. On the other end of the spectrum, we have [benchmarks](https://github.com/json-api-dotnet/JsonApiDotNetCore/tree/master/benchmarks) using [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) to measure the performance and allocations of a small piece of code.

What we don't have is an easy way to see how much time is spent (relatively and absolutely) inside the various layers (controllers, services, repositories, serialization, etc) of JADNC. It would enable us to identify code paths worth considering for optimization. And seeing the overall performance impact on changes like switching from Newtonsoft.Json to System.Text.Json.

### PR DESCRIPTION
**Pull request by bart-degreed**
This PR instruments the various layers of code in the library and the example project. It enables identification of code paths worth considering for optimization.

Fixes #1032.

Example output:
```
JsonApiDotNetCoreExample.Startups.Startup: Information: Measurement results for application startup:
Configure other (startup) .........  0:00:00:00.0018732 ...   0.17%
  Configure JSON:API (startup) ....  0:00:00:01.0907241 ...  99.83%
Initialize other (startup) ........  0:00:00:00.5180028 ...  98.60%
  Initialize JSON:API (startup) ...  0:00:00:00.0073801 ...   1.40%

JsonApiDotNetCore.Middleware.JsonApiMiddleware: Information: Measurement results for GET http://localhost:14140/api/v1/todoItems?include=owner,assignee,tags:
JSON:API middleware .................................  0:00:00:00.0000286 ...   1.14%
  Subsequent middleware .............................  0:00:00:00.0002976 ...  11.86%
    Parse query string ..............................  0:00:00:00.0000360 ...   1.43%
    Service - Get resources .........................  0:00:00:00.0001394 ...   5.55%
      Repository - Count resources ..................  0:00:00:00.0000030 ...   0.12%
        Convert QueryLayer to System.Expression .....  0:00:00:00.0001265 ...   5.04%
          Convert System.Expression to IQueryable ...  0:00:00:00.0000009 ...   0.04%
        Execute SQL (count) .........................  0:00:00:00.0122070
      Top-level query composition ...................  0:00:00:00.0000984 ...   3.92%
      Nested query composition ......................  0:00:00:00.0000351 ...   1.40%
      Repository - Get resource(s) ..................  0:00:00:00.0000021 ...   0.08%
        Convert QueryLayer to System.Expression .....  0:00:00:00.0003258 ...  12.98%
          Convert System.Expression to IQueryable ...  0:00:00:00.0000013 ...   0.05%
        Execute SQL (data) ..........................  0:00:00:00.0124639
    Write response body .............................  0:00:00:00.0004246 ...  16.92%
      Serializer.Build (list) .......................  0:00:00:00.0007376 ...  29.39%
      Newtonsoft.Serialize ..........................  0:00:00:00.0002527 ...  10.07%
```

Note that the percentages in non-leaves indicate the relative time spent _in that call_, they are not cumulative. For example, in the request above, the relative time spent in "Write response body" is 16.92% (self) + 29.39% + 10.07% (children) = ~55%.

### CODE REVIEW

**Comment by bart-degreed** [comment]
Reverted Resharper/Rider update due to hanging cibuild. Tracked at https://youtrack.jetbrains.com/issue/RSRP-481371#focus=Comments-27-5106579.0-0.

File: src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // Be default, measurements using Stopwatch can differ 25%-30% on the same function on the same computer.
+                // The steps below ensure to get an accuracy of 0.1%-0.2%. With this accuracy, algorithms can be tested and compared.
+                // https://www.codeproject.com/Articles/61964/Performance-Tests-Precise-Run-Time-Measurements-wi
+
+                // The most important thing is to prevent switching between CPU cores or processors. Switching dismisses the cache, etc. and has a huge performance impact on the test.
+                Process.GetCurrentProcess().ProcessorAffinity = new IntPtr(2);
+
+                // To get the CPU core more exclusively, we must prevent that other threads can use this CPU core. We set our process and thread priority to achieve this.
```
Comment (maurei) [inline src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs:31]: We set our process ~and thread~ priority to achieve this.

File: src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    /// </summary>
+    internal sealed class CascadingCodeTimer : ICodeTimer
+    {
+        private readonly Stopwatch _stopwatch = new();
+        private readonly Stack<MeasureScope> _activeScopeStack = new();
+        private readonly List<MeasureScope> _completedScopes = new();
+
+        static CascadingCodeTimer()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (maurei) [inline src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs:22]: if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))

**Review by maurei — CHANGES_REQUESTED** [review]
Added two comments earlier

File: src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs (hunk 2 above)
Comment (bart-degreed) [reply src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs:22]: While not failing on OSX, setting this does not actually help. See https://github.com/dotnet/runtime/blob/defa26b9e1159f1a2a3470c52084410edff982b6/src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs#L931.

File: src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs (hunk 1 above)
Comment (bart-degreed) [reply src/JsonApiDotNetCore/Diagnostics/CascadingCodeTimer.cs]: Updated.

**Review by maurei — APPROVED** [review]

**Pull request merged**